pkg/kubelet/kubeletconfig/controller.go

```diff
@@ -0,0 +1,350 @@
+	pendingConfigSource chan bool
+
+	// configOK manages the ConfigOK condition that is reported in Node.Status.Conditions
+	configOK status.ConfigOKCondition
+
+	// informer is the informer that watches the Node object
+	informer cache.SharedInformer
+
+	// checkpointStore persists config source checkpoints to a storage layer
+	checkpointStore store.Store
+
+	// badConfigTracker persists bad-config records to a storage layer
+	badConfigTracker badconfig.Tracker
+
+	// startupTracker persists Kubelet startup records, used for crash-loop detection, to a storage layer
+	startupTracker startups.Tracker
+}
+
+// NewController constructs a new Controller object and returns it. Directory paths must be absolute.
+// If the `initConfigDir` is an empty string, skips trying to load the init config.
+// If the `dynamicConfigDir` is an empty string, skips trying to load checkpoints or download new config,
+// but will still sync the ConfigOK condition if you call StartSync with a non-nil client.
+func NewController(initConfigDir string, dynamicConfigDir string, defaultConfig *componentconfig.KubeletConfiguration) *Controller {
+	fs := utilfs.DefaultFs{}
+
+	var initLoader configfiles.Loader
+	if len(initConfigDir) > 0 {
+		initLoader = configfiles.NewFSLoader(fs, initConfigDir)
+	}
+	dynamicConfig := false
+	if len(dynamicConfigDir) > 0 {
+		dynamicConfig = true
+	}
+
+	// Get the current kubelet version; bad-config and startup-tracking information can be kubelet-version specific,
+	// e.g. a bug that crash loops an old Kubelet under a given config might be fixed in a new Kubelet or vice-versa,
+	// validation might be relaxed in a new Kubelet, etc.
+	// We also don't want a change in a file format to break Kubelet upgrades; this makes sure a new kubelet gets
+	// a fresh dir to put its config health data in.
+	// Note that config checkpoints use the api machinery to store ConfigMaps, and thus get file format versioning for free.
+	kubeletVersion := version.Get().String()
+
+	return &Controller{
+		dynamicConfig: dynamicConfig,
+		defaultConfig: defaultConfig,
+		// channels must have capacity at least 1, since we signal with non-blocking writes
+		pendingConfigSource: make(chan bool, 1),
+		configOK:            status.NewConfigOKCondition(),
+		checkpointStore:     store.NewFsStore(fs, filepath.Join(dynamicConfigDir, checkpointsDir)),
+		badConfigTracker:    badconfig.NewFsTracker(fs, filepath.Join(dynamicConfigDir, badConfigTrackingDir, kubeletVersion)),
+		startupTracker:      startups.NewFsTracker(fs, filepath.Join(dynamicConfigDir, startupTrackingDir, kubeletVersion)),
+		initLoader:          initLoader,
+	}
+}
+
+// Bootstrap attempts to return a valid KubeletConfiguration based on the configuration of the Controller,
+// or returns an error if no valid configuration could be produced. Bootstrap should be called synchronously before StartSync.
+func (cc *Controller) Bootstrap() (*componentconfig.KubeletConfiguration, error) {
```

Comment:
(2) is covered by existing e2e_node tests

(1, 4, 5) were covered by the [manual test cases](https://docs.google.com/document/d/1L_bw_Dy4QgEOCSAIO_oxM8GxJVkWz84omZFLV-SWuk8/edit#), all of which I plan to turn into automated tests after this PR merges - but I'm trying not to increase the size of this PR any further

(3) we need to figure out how to test this. I don't want to rely on a "valid" config that causes a crash loop - because the very existence of such a config should be viewed as a bug. I also don't want to include a "please crash loop for our tests" field in the config, because that's an ugly API. The best solution is not clear here.